File: QUESTION FETCHER ENGINE/Manya-app-master/migrate_to_pg.py
import sqlite3
import psycopg2

# --- Configuration ---
SQLITE_PATH = "manya.db"
PG_CONN_STRING = "host=localhost port=5432 dbname=manya_db user=postgres password=postgres"

# --- Connect to both databases ---
sqlite_conn = sqlite3.connect(SQLITE_PATH)
sqlite_cursor = sqlite_conn.cursor()

pg_conn = psycopg2.connect(PG_CONN_STRING)
pg_cursor = pg_conn.cursor()
pg_conn.autocommit = False  # Use transactions for safety

# --- Get list of tables from SQLite ---
sqlite_cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
tables = [table[0] for table in sqlite_cursor.fetchall()]

print(f"Found tables: {tables}")

for table_name in tables:
    print(f"\n--- Migrating {table_name} ---")
    
    # 1. Get table schema
    sqlite_cursor.execute(f"SELECT sql FROM sqlite_master WHERE type='table' AND name='{table_name}';")
    create_sql = sqlite_cursor.fetchone()[0]
    
    # 2. Convert SQLite schema to PostgreSQL
    #    - Replace AUTOINCREMENT with SERIAL
    #    - Remove sqlite_sequence references
    create_sql = create_sql.replace("INTEGER PRIMARY KEY AUTOINCREMENT", "SERIAL PRIMARY KEY")
    create_sql = create_sql.replace("AUTOINCREMENT", "")
    
    # 3. Drop and recreate table in PostgreSQL
    pg_cursor.execute(f"DROP TABLE IF EXISTS {table_name} CASCADE;")
    pg_cursor.execute(create_sql)
    print(f"   Table created.")
    
    # 4. Get data from SQLite
    sqlite_cursor.execute(f"SELECT * FROM {table_name};")
    rows = sqlite_cursor.fetchall()
    
    if rows:
        # Get column names (excluding the sqlite_sequence table which is internal)
        if table_name != "sqlite_sequence":
            sqlite_cursor.execute(f"PRAGMA table_info({table_name});")
            columns = [col[1] for col in sqlite_cursor.fetchall()]
            placeholders = ','.join(['%s'] * len(columns))
            insert_sql = f"INSERT INTO {table_name} ({','.join(columns)}) VALUES ({placeholders})"
            
            # Insert data in batches
            batch_size = 100
            for i in range(0, len(rows), batch_size):
                batch = rows[i:i+batch_size]
                pg_cursor.executemany(insert_sql, batch)
                pg_conn.commit()
                print(f"   Inserted rows {i} to {i+len(batch)}")
    
    pg_conn.commit()

print("\n✅ Migration complete!")

# --- Clean up ---
sqlite_conn.close()
pg_cursor.close()
pg_conn.close()

File: Manya-app-master (1)/Manya-app-master/migrate-to-pg.js
const sqlite3 = require('sqlite3');
const { Pool } = require('pg');
const fs = require('fs');

// Configuration
const SQLITE_PATH = './manya.db';
const PG_CONFIG = {
    host: 'localhost',
    port: 5432,
    database: 'manya_db',
    user: 'postgres',
    password: 'root', // Change this to your actual password
};

async function migrate() {
    console.log('🔄 Starting migration from SQLite to PostgreSQL...');
    
    // Connect to SQLite
    const sqlite = new sqlite3.Database(SQLITE_PATH, (err) => {
        if (err) {
            console.error('❌ Error connecting to SQLite:', err.message);
            process.exit(1);
        }
        console.log('✅ Connected to SQLite');
    });

    // Connect to PostgreSQL
    const pg = new Pool(PG_CONFIG);
    
    try {
        // Test PostgreSQL connection
        await pg.query('SELECT 1');
        console.log('✅ Connected to PostgreSQL');
    } catch (err) {
        console.error('❌ Error connecting to PostgreSQL:', err.message);
        process.exit(1);
    }

    // Get all tables from SQLite
    const tables = await new Promise((resolve, reject) => {
        sqlite.all(
            "SELECT name FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%'",
            [],
            (err, rows) => {
                if (err) reject(err);
                else resolve(rows.map(r => r.name));
            }
        );
    });

    console.log('\n📋 Found tables in SQLite:', tables);

    // Migrate each table
    for (const table of tables) {
        console.log(`\n📦 Migrating table: ${table}`);
        
        try {
            // Get table schema
            const schema = await new Promise((resolve, reject) => {
                sqlite.all(`PRAGMA table_info(${table})`, [], (err, rows) => {
                    if (err) reject(err);
                    else resolve(rows);
                });
            });

            const columns = schema.map(col => col.name);
            console.log(`   Columns: ${columns.join(', ')}`);

            // Drop table if exists in PostgreSQL
            try {
                await pg.query(`DROP TABLE IF EXISTS "${table}" CASCADE`);
                console.log(`   Dropped existing table`);
            } catch (err) {
                // Ignore drop errors
            }

            // Create table in PostgreSQL
            const createColumns = schema.map(col => {
                let type = col.type.toUpperCase();
                // Convert SQLite types to PostgreSQL
                if (type.includes('INT')) {
                    if (col.pk) return `"${col.name}" SERIAL PRIMARY KEY`;
                    return `"${col.name}" INTEGER`;
                }
                if (type.includes('CHAR') || type.includes('TEXT')) return `"${col.name}" TEXT`;
                if (type.includes('REAL') || type.includes('DOUB')) return `"${col.name}" DOUBLE PRECISION`;
                if (type.includes('BOOL')) return `"${col.name}" BOOLEAN`;
                if (type.includes('DATE') || type.includes('TIME')) return `"${col.name}" TIMESTAMP`;
                return `"${col.name}" TEXT`;
            }).join(',\n            ');

            const createSQL = `CREATE TABLE "${table}" (\n            ${createColumns}\n        )`;
            
            try {
                await pg.query(createSQL);
                console.log(`   ✅ Table created`);
            } catch (err) {
                console.error(`   ❌ Error creating table:`, err.message);
                continue;
            }

            // Get data from SQLite
            const rows = await new Promise((resolve, reject) => {
                sqlite.all(`SELECT * FROM "${table}"`, [], (err, rows) => {
                    if (err) reject(err);
                    else resolve(rows);
                });
            });

            if (rows.length === 0) {
                console.log(`   📭 No data to migrate`);
                continue;
            }

            // Insert data in batches
            const batchSize = 100;
            let inserted = 0;
            
            for (let i = 0; i < rows.length; i += batchSize) {
                const batch = rows.slice(i, i + batchSize);
                
                // Build parameterized insert query
                const placeholders = batch.map((_, idx) => 
                    `(${columns.map((_, j) => `$${idx * columns.length + j + 1}`).join(', ')})`
                ).join(', ');
                
                const values = batch.flatMap(row => 
                    columns.map(col => {
                        const val = row[col];
                        // Handle SQLite-specific values
                        if (val === null) return null;
                        if (typeof val === 'object') return JSON.stringify(val);
                        return val;
                    })
                );

                const insertSQL = `INSERT INTO "${table}" (${columns.map(c => `"${c}"`).join(', ')}) VALUES ${placeholders}`;

                try {
                    await pg.query(insertSQL, values);
                    inserted += batch.length;
                    console.log(`   ✅ Inserted rows ${i + 1} to ${i + batch.length} (${inserted}/${rows.length})`);
                } catch (err) {
                    console.error(`   ❌ Error inserting batch:`, err.message);
                    // Try one by one for debugging
                    console.log(`   🔍 Attempting single-row inserts for debugging...`);
                    for (const row of batch) {
                        const singleValues = columns.map(col => row[col]);
                        const singlePlaceholders = columns.map((_, j) => `$${j + 1}`).join(', ');
                        const singleSQL = `INSERT INTO "${table}" (${columns.map(c => `"${c}"`).join(', ')}) VALUES (${singlePlaceholders})`;
                        try {
                            await pg.query(singleSQL, singleValues);
                        } catch (singleErr) {
                            console.error(`      ❌ Failed on row:`, row);
                            console.error(`      Error:`, singleErr.message);
                        }
                    }
                }
            }

            console.log(`   ✅ Migrated ${inserted} rows to ${table}`);

        } catch (err) {
            console.error(`   ❌ Error migrating table ${table}:`, err.message);
        }
    }

    console.log('\n🎉 Migration complete!');
    
    // Close connections
    sqlite.close();
    await pg.end();
}

// Run the migration
migrate().catch(err => {
    console.error('Migration failed:', err);
    process.exit(1);
});